pkg/binding/binding.go

```diff
@@ -146,3 +155,184 @@ func (o *BindingClient) GetServiceInstances() (map[string]unstructured.Unstructu
+		strCRD, err := libdevfile.GetK8sManifestWithVariablesSubstituted(devfileObj, component.Name, context, devfilefs.DefaultFs{})
+		if err != nil {
+			return nil, err
+		}
+
+		u := unstructured.Unstructured{}
+		if err := yaml.Unmarshal([]byte(strCRD), &u.Object); err != nil {
+			return nil, err
+		}
+
+		switch u.GetObjectKind().GroupVersionKind() {
+		case bindingApi.GroupVersionKind:
+
+			var sbo bindingApi.ServiceBinding
+			err := o.kubernetesClient.ConvertUnstructuredToResource(u, &sbo)
+			if err != nil {
+				return nil, err
+			}
+
+			sb, err := api.ServiceBindingFromBinding(sbo)
+			if err != nil {
+				return nil, err
+			}
+
+			sb.Status, err = o.getStatusFromBinding(sb.Name)
+			if err != nil {
+				return nil, err
+			}
+
+			result = append(result, sb)
+
+		case specApi.GroupVersion.WithKind("ServiceBinding"):
+
+			var sbc specApi.ServiceBinding
+			err := o.kubernetesClient.ConvertUnstructuredToResource(u, &sbc)
+			if err != nil {
+				return nil, err
+			}
+
+			sb, err := api.ServiceBindingFromSpec(sbc)
+			if err != nil {
+				return nil, err
+			}
+
+			sb.Status, err = o.getStatusFromSpec(sb.Name)
+			if err != nil {
+				return nil, err
+			}
+
+			result = append(result, sb)
+
+		}
+	}
+	return result, nil
+}
+
+// GetBinding returns the ServiceBinding retource with the given name
+// from the cluster, from group binding.operators.coreos.com/v1alpha1 or servicebinding.io/v1alpha3
+func (o *BindingClient) GetBinding(name string) (api.ServiceBinding, error) {
+
+	bindingSB, err := o.kubernetesClient.GetBindingServiceBinding(name)
+	if err == nil {
+		var sb api.ServiceBinding
+		sb, err = api.ServiceBindingFromBinding(bindingSB)
+		if err != nil {
+			return api.ServiceBinding{}, err
+		}
+		sb.Status, err = o.getStatusFromBinding(bindingSB.Name)
+		if err != nil {
+			return api.ServiceBinding{}, err
+		}
+		return sb, nil
+	}
+	if err != nil && !kerrors.IsNotFound(err) {
+		return api.ServiceBinding{}, err
+	}
+
+	specSB, err := o.kubernetesClient.GetSpecServiceBinding(name)
+	if err == nil {
+		var sb api.ServiceBinding
+		sb, err = api.ServiceBindingFromSpec(specSB)
+		if err != nil {
+			return api.ServiceBinding{}, err
+		}
+		sb.Status, err = o.getStatusFromSpec(specSB.Name)
+		if err != nil {
+			return api.ServiceBinding{}, err
+		}
+		return sb, nil
+	}
+
+	// In case of notFound error, this time we return the error
+	if kerrors.IsNotFound(err) {
+		return api.ServiceBinding{}, fmt.Errorf("ServiceBinding %q not found", name)
+	}
+	return api.ServiceBinding{}, err
+}
+
+// getStatusFromBinding returns status information from a ServiceBinding in the cluster
+// from group binding.operators.coreos.com/v1alpha1
+func (o *BindingClient) getStatusFromBinding(name string) (*api.ServiceBindingStatus, error) {
+	bindingSB, err := o.kubernetesClient.GetBindingServiceBinding(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	secretName := bindingSB.Status.Secret
```

Comment:
Putting it here for the record...

As discussed, `secretName` might be empty if the ServiceBinding instance is in a non-ready state. As seen together, the ServiceBinding status needs to be checked before reading the secret.

Otherwise, this might cause issues like this when reading the secret afterwards:

```
❯ odo describe binding
 ✗  unable to get the secret &Secret{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Data:map[string][]byte{},Type:,StringData:map[string]string{},Immutable:nil,}: resource name may not be empty
```

This happened after binding to a Redis instance, added via the [Redis Operator](https://operatorhub.io/operator/redis-operator) and the Redis CRD:
```
❯ kubectl get servicebindings my-miniflux-app-redis-standalone
NAME                               READY   REASON            AGE
my-miniflux-app-redis-standalone   False   ProcessingError   149m

❯ kubectl get servicebindings my-miniflux-app-redis-standalone -o=jsonpath='{.status}' | jq   
{
  "conditions": [
    {
      "lastTransitionTime": "2022-05-31T11:36:09Z",
      "message": "redisSecret is not found",
      "reason": "ErrorReadingBinding",
      "status": "False",
      "type": "CollectionReady"
    },
    {
      "lastTransitionTime": "2022-05-31T11:36:09Z",
      "message": "redisSecret is not found",
      "reason": "ProcessingError",
      "status": "False",
      "type": "Ready"
    }
  ],
  "secret": ""
}
```

Reply:
With hindsight, shouldn't we check that the secret does exist as well (since we are getting the binding information from the secret)? I just manually deleted the secret in my local cluster (thinking it would maybe get recreated, but it was not). Then I hit the same issue again:

```
❯ odo describe binding
 ✗  unable to get the secret &Secret{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Data:map[string][]byte{},Type:,StringData:map[string]string{},Immutable:nil,}: secrets "my-miniflux-postgresql-db-28d34d55" not found
```

I think this should allow us to write a test case for this, if the secret is missing, regardless of the service binding status.